src/impl/tpls/KokkosBlas3_gemm_tpl_spec_decl.hpp

```diff
@@ -335,6 +335,127 @@ KOKKOSBLAS3_CGEMM_BLAS( Kokkos::LayoutRight, Kokkos::LayoutRight, Kokkos::Layout
 namespace KokkosBlas {
 namespace Impl {
 
+
+// DotBasedGEMM implements the optimization for C = beta*C + alpha*A^TB 
+// with A and B matrices both being tall and skinny. C matrix is assumably 
+// small, so, each entry of C is computed by performing the dot product of 
+// respective columns of A and B matrices. Note that the dot products are
+// performed on very long vectors, so, each dot product is distributed among
+// numDivPerDot teams.     
+
+struct TagInit{};   // The tag for the initialization parallel_for 
+struct TagMult{};   // The tag for the multiplication parallel_for 
+template<class ExecSpace, class AV, class BV, class CV>
+struct DotBasedGEMM{
+
+  AV A;
+  BV B;
+  CV C;
+
+  typedef typename AV::non_const_value_type scalar_A;
```

Comment:
Prefer C++11 `using` syntax in future commits.

Reply:
Done.